### PR TITLE
Use `partial_cmp` to implement tuple `lt`/`le`/`ge`/`gt`

### PR DESCRIPTION
**Pull request by scottmcm**
In today's implementation, `(A, B)::gt` contains calls to *both* `A::eq` *and* `A::gt`.

That's fine for primitives, but for things like `String`s it's kinda weird -- `(String, usize)::gt` has a call to both `bcmp` and `memcmp` (<https://rust.godbolt.org/z/7jbbPMesf>) because when `bcmp` says the `String`s aren't equal, it turns around and calls `memcmp` to find out which one's bigger.

This PR changes the implementation to instead implement `(A, …, C, Z)::gt` using `A::partial_cmp`, `…::partial_cmp`, `C::partial_cmp`, and `Z::gt`.  (And analogously for `lt`, `le`, and `ge`.)  That way expensive comparisons don't need to be repeated.

Technically this is an observable change on stable, so I've marked it `needs-fcp` + `T-libs-api` and will
r? rust-lang/libs-api 

I'm hoping that this will be non-controversial, however, since it's very similar to the observable changes that were made to the derives (#81384 #98655) -- like those, this only changes behaviour if a type overrode behaviour in a way inconsistent with the rules for the various traits involved.

(The first commit here is #108156, adding the codegen test, which I used to make sure this doesn't regress behaviour for primitives.)

Zulip conversation about this change: <https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/.60.3E.60.20on.20Tuples/near/328392927>.

### CODE REVIEW

File: library/core/benches/tuple.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        ("core::iter::adapters::Copie", 789_usize),
+        ("core::iter::adapters::Cycle", 123_usize),
+        ("core::iter::adapters::Flatt", 456_usize),
+        ("core::iter::adapters::TakeN", 789_usize),
+    ]);
+
+    b.iter(|| {
+        let x = data.choose(&mut rng).unwrap();
+        let y = data.choose(&mut rng).unwrap();
+        [x < y, x <= y, x > y, x >= y]
```
Comment (scottmcm) [inline library/core/benches/tuple.rs:20]: As requested on Zulip, a quick bench to show the difference, though of course different-length strings wouldn't show as much of a difference.

Before:
```
running 1 test
test tuple::bench_tuple_comparison                 ... bench:      63 ns/iter (+/- 1)
```

After
```
running 1 test
test tuple::bench_tuple_comparison                 ... bench:      26 ns/iter (+/- 0)
```

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by scottmcm** [comment]
**tm** pointed out on zulip that the previous implementation was in fact quadratic in the nesting level.  **Mario Carneiro** produced a [demonstration](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=017a04f60fd7861c9e4cabdbf8434b33) of that, so with sufficient nesting this PR (which is linear instead) can have an arbitrarily-high performance improvement 🙃

**Review by dtolnay — APPROVED** [review]

**Pull request merged**